### PR TITLE
Stricter validation for integer and float configuration

### PR DESCRIPTION
**Pull request by marcotc**
**2.0 Upgrade Guide notes**
Configuration options of types ':int' and ':float' now have stricter type validation:
* Fractional numbers are no longer considered valid for ':int' options.
* Empty values (nil or empty string) are no longer considered valid for ':int' or ':float' options.

This applies to both environment variable and programmatic configuration.
Here's the list of all affected options:
[Insert all :int and :float options at release time!!!]

The environment variable `DD_EXPERIMENTAL_SKIP_CONFIGURATION_VALIDATION` has been removed.

**Motivation:**
It was possible to configure options with empty values (e.g. `DD_TRACE_SAMPLE_RATE=`) and end up with a sampling rate of `0.0`, because of the logic `value.to_f`. The same applies to integer options (`value.to_i`).
Non-numeric values would also be converted to zero.
This means users do not know what their effective configuration will be, and are not given feedback on how to address configuration issues.

**How to test the change?**
There are unit tests for all the changes.

**For Datadog employees:**
- [ ] If this PR touches code that signs or publishes builds or packages, or handles
credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: lib/datadog/core/configuration/option.rb [hunk 1]
```diff
@@ -191,11 +189,33 @@ def coerce_env_variable(value)
...
+        def coerce_env_var_int(value)
+          # Strict integer parsing
+          Integer(value)
+        rescue ArgumentError => e
+          # Try also parsing as a whole floating point numbers
+          begin
+            f = Float(value)
+
+            # Check for whole numbers
+            raise ArgumentError, "#{value} is not a whole number" unless f.truncate == f
```
Comment (ivoanjo) [inline lib/datadog/core/configuration/option.rb:206]: To be honest... This seems like a bit of a surprising semantics. If I'm in a rush and see something on my config that I want to update:

```
DD_PROFILING_HOW_MUCH_IT_ROCKS=100.0
```

and then I go, oh, I want to make it rock a bit less

```
DD_PROFILING_HOW_MUCH_IT_ROCKS=95.5
```

and it blows up -- this seems weird. That .0 that someone on my team left misled me into thinking this value was supposed to be a float, and then it turns out it isn't and it doesn't work.

File: lib/datadog/core/configuration/option.rb [hunk 2]
```diff
@@ -191,11 +189,33 @@ def coerce_env_variable(value)
...
           else
-            raise ArgumentError,
+            raise InvalidDefinitionError,
               "The option #{@definition.name} is using an unsupported type option for env coercion `#{@definition.type}`"
           end
         end
 
+        def coerce_env_var_int(value)
+          # Strict integer parsing
+          Integer(value)
```
Comment (ivoanjo) [inline lib/datadog/core/configuration/option.rb:199]: I learned (depressingly) the other day that `Integer(value)` actually supports hex and octal literals => e.g. `Integer("010") => 8` and `Integer("0x10") => 16`.

TL;DR the correct form is `Integer(value, 10)` to specifically ask it to parse only as base 10. Now I'll go back to wondering why this isn't the default >_>

**Review by ivoanjo — COMMENTED** [review]
I'm a bit short on time today so I didn't look at this very thoroughly but here's a quick pass :)

File: lib/datadog/core/configuration/option.rb (hunk 1 above)
Comment (marcotc) [reply lib/datadog/core/configuration/option.rb:206]: What do you think about allowing whole floats as integer @TonyCTHsu @ekump? (e.g. `100.0` is allowed for `:int` configuration options). Or should be strict and say that this is not allowed (e.g. `100.0` would be rejected, and the user should replace it with an explicit `100`)

File: lib/datadog/core/configuration/option.rb (hunk 2 above)
Comment (marcotc) [reply lib/datadog/core/configuration/option.rb:199]: I think if only `0x10` (and base 10) were allowed, we wouldn't bat an eye, but octals ruin it for everyone.

Comment (marcotc) [reply lib/datadog/core/configuration/option.rb:199]: Fixed it!

File: lib/datadog/core/configuration/option.rb (hunk 1 above)
Comment (marcotc) [reply lib/datadog/core/configuration/option.rb:206]: @ivoanjo, how about a warning on whole floats? I think that might be a good compromise.

Comment (ivoanjo) [reply lib/datadog/core/configuration/option.rb:206]: > how about a warning on whole floats?

I would still kinda prefer not to accept 'em, but I can live with that as a compromise ;)

File: lib/datadog/core/configuration/option.rb (hunk 2 above)
Comment (ivoanjo) [reply lib/datadog/core/configuration/option.rb]: Yeah, it's hard to accidentally have hex, but so easy to accidentally have octal >_>

File: lib/datadog/core/configuration/option.rb (hunk 1 above)
Comment (ivoanjo) [reply lib/datadog/core/configuration/option.rb:206]: We had discussed moving this to a warning, which I think is better than the current raise. If we do keep the raise, I would suggest making it a bit clear what needs to be fixed in the current error -- e.g. "Setting expects only integer settings, you've provided #{value}"?

File: lib/datadog/core/configuration/option.rb [hunk 3]
```diff
@@ -191,11 +189,33 @@ def coerce_env_variable(value)
...
+            f = Float(value)
+
+            # Check for whole numbers
+            raise ArgumentError, "#{value} is not a whole number" unless f.truncate == f
+
+            f
+          rescue ArgumentError
+            # It's neither an integer nor a whole float
+            raise e
+          end
```
Comment (ivoanjo) [inline lib/datadog/core/configuration/option.rb:212]: Is it just me, or would this `rescue` swallow our `raise ArgumentError, "#{value} is not a whole number" unless f.truncate == f` above and replace it with the original `Integer` parsing error?

(I guess it may be worth adding a spec to validate that the right error/message gets presented?)

File: lib/datadog/core/configuration/option.rb [hunk 4]
```diff
@@ -191,11 +189,33 @@ def coerce_env_variable(value)
...
+          Integer(value, 10)
+        rescue ArgumentError => e
+          # Try also parsing as a whole floating point numbers
+          begin
+            f = Float(value)
+
+            # Check for whole numbers
+            raise ArgumentError, "#{value} is not a whole number" unless f.truncate == f
+
+            f
```
Comment (ivoanjo) [inline lib/datadog/core/configuration/option.rb:208]: Minor: Output an `Integer` in the end?

```suggestion
            f.to_i
```

File: lib/datadog/core/configuration/option.rb [hunk 5]
```diff
@@ -235,8 +255,10 @@ def validate(type, value)
           case type
           when :string
             value.is_a?(String)
-          when :int, :float
-            value.is_a?(Numeric)
+          when :int
+            value.is_a?(Integer) || (value.is_a?(Float) && (value.truncate == value))
+          when :float
+            value.is_a?(Float) || value.is_a?(Integer) || value.is_a?(Rational)
```
Comment (ivoanjo) [inline lib/datadog/core/configuration/option.rb:261]: This makes me think we should change `validate_type` above to coerce int and float to the correct types.

It seems like a bit of a potential sharp edge that someone can set a `Rational` on a `:float` setting, or a `Float` on `:int` but our code may not deal very well with it.

On the other hand, it's nice to allow these; hence why I think here would be a nice place to coerce after validation?

**Review by ivoanjo — COMMENTED** [review]
Left a few final notes :)

**Comment by marcotc** [comment]
@ivoanjo I talked to @TonyCTHsu and I was convinced to have stricter Integer validation, which aligns with your original suggestion.

File: lib/datadog/core/configuration/option.rb [hunk 6]
```diff
@@ -235,8 +232,10 @@ def validate(type, value)
           case type
           when :string
             value.is_a?(String)
-          when :int, :float
-            value.is_a?(Numeric)
+          when :int
+            value.is_a?(Integer)
+          when :float
+            value.is_a?(Float) || value.is_a?(Rational)
```
Comment (ivoanjo) [inline lib/datadog/core/configuration/option.rb:238]: I think allowing users to provide integer values for floats is a nice UX touch -- e.g. if you're providing a percentage, often you may want to say 50 and not 50.0 or something like that.

So I don't think we should lose that.

**Review by ivoanjo — APPROVED** [review]
I think we should still allow users to provide `Integer` to float settings when doing it via the configure block, but otherwise this looks good to me :+1:

**Review by ivoanjo — APPROVED** [review]
:+1: LGTM
